src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/Tensor.cs

```diff
@@ -0,0 +1,355 @@
+            _strides = SpanHelpers.CalculateStrides(Rank, _lengths);
+            _isPinned = isPinned;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal Tensor(T[] values, nint[] lengths, nint[] strides, bool isPinned)
+        {
+            _linearLength = SpanHelpers.CalculateTotalLength(ref lengths);
+
+            Debug.Assert(_linearLength >= 0);
+
+            _values = values;
+            _lengths = lengths;
+            _strides = strides;
+            if (strides == Array.Empty<nint>())
+                _strides = SpanHelpers.CalculateStrides(Rank, lengths);
+            _isPinned = isPinned;
+
+        }
+
+        // IDisposable
+
+        public void Dispose()
+        {
+        }
+
+        private static class EmptyTensor
+        {
+#pragma warning disable CA1825 // this is the implementation of Tensor.Empty<T>()
+            internal static readonly Tensor<T> Value = new Tensor<T>();
+#pragma warning restore CA1825
+        }
+        // ITensor
+
+        public static Tensor<T> Empty => EmptyTensor.Value;
+
+        public bool IsEmpty { get { return _lengths.Length == 0; } }
+
+        public bool IsPinned { get { return _isPinned; } }
+
+        public int Rank { get { return _lengths.Length; } }
+
+        /// <summary>
+        /// The number of items in the span.
+        /// </summary>
+        internal nint LinearLength
+        {
+            get => _linearLength;
+        }
+        // REIVEW: SHOULD WE CALL THIS DIMS OR DIMENSIONS?
+        /// <summary>
+        /// Gets the length of each dimension in this <see cref="Tensor{T}"/>.
+        /// </summary>
+        public ReadOnlySpan<nint> Lengths { get { return _lengths; } }
+        public ReadOnlySpan<nint> Strides { get { return _strides; } }
+
+        /// <summary>
+        /// Returns a reference to specified element of the Span.
+        /// </summary>
+        /// <param name="indices"></param>
+        /// <returns></returns>
+        /// <exception cref="IndexOutOfRangeException">
+        /// Thrown when index less than 0 or index greater than or equal to Length
+        /// </exception>
+        public ref T this[params nint[] indices] => ref this[indices.AsSpan()];
+
+        /// <summary>
+        /// Returns a reference to specified element of the Span.
+        /// </summary>
+        /// <param name="indices"></param>
+        /// <returns></returns>
+        /// <exception cref="IndexOutOfRangeException">
+        /// Thrown when index less than 0 or index greater than or equal to Length
+        /// </exception>
+        public ref T this[ReadOnlySpan<nint> indices] => ref AsSpan()[indices];
+
+        // REVIEW: NOT IN API DOC BUT IN NOTEBOOK AND IN OTHER FRAMEWORKS
+        // REVIEW: WE WILL WANT THIS CHANGED FROM A BOOL TO SOME FILTER EXPRESSION.
+        public Tensor<T> this[Tensor<bool> filter]
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
Isn't this a sizeable amount of code to aggressively inline? It's doing quite a lot of work... is the inlining actually meaningful?

Reply:
It is. That was a copy paste error on my part. Removed.